### ISSUE TITLE
Confusion in function update_routing

### ISSUE THREAD

**Issue by HopefulRational** [brjathu/deepcaps#4]
@brjathu Hi. I've read your paper and it's great that you've given your code. Thanks a lot. I have a few doubts in the `ConvCapsuleLayer3D` layer. Could you clarify them for me?
1. In the paper, the `softmax_3D` function was given, with the denominator summed over spatial positions and the capsules in the l+1 th layer. The implementation, however, uses `nn.softmax` whose denominator sums over only the capsules in `update_routing`. Could you clarify this?
2. In the `update_routing` function, I am unable to understand the reshaping operations. It seems that the dimension for the batch size is being changed. Could you clarify the sizes used?
Thank you.

**Comment by HopefulRational** [comment]
I have read your paper. In the paper, the softmax3d function was given, and the denominator summed over spatial positions and the capsules in the l+1 th layer. The implementation, however, uses `nn.softmax` in `update_routing` whose denominator sums over only the capsules. Could you clarify this?

**Comment by brjathu** [comment]
@HopefulRational Hi, Thanks for your interest in our work.
1. Our apologies for making the code, and figs bit complex. In the paper the `l+1` means the input to the `l+1`th layer, which is actually the output from layer `l` just after routing. Further, since the routing is iterative, this softmax is applied inside `update_routing` function on the output of the routing at each iteration.
2.  Yes, I accept that the reshaping is too complicated. Yes, the `batch_size` dimension is changed to keep the multiplication with `logits`, without keeping the number of atoms in the dimension. but `votes_t_shape` and `r_t_shape` are simply transpose of themselves, and `r_t_shape` used later convert the votes back to early shape.

Again we are sorry for all the trouble you have to go through to understand the code, we will be releasing the pytorch implementation soon, which would be much easier to understand.

Please feel free to ask any questions or concerns you have.

Thank you very much.

**Comment by HopefulRational** [comment]
Hi, thank you for your quick reply. I am waiting for the pytorch release!
While I understand the shapes (somewhat) in the convcapsulelayer3d class routing, I still do not see how the softmax in the code and the softmax3d function in the paper are the same.
tf.nn.softmax() is not the same as softmax3d, as softmax3d sums over spatial positions and capsules, while tf.nn.softmax() sums only over the capsules. This is in the function `update_routing`.
Could you clarify, please?

**Comment by brjathu** [comment]
Hi, 

I see your point, Let me check with other authors and compare it with our the base code and I will get back to you with correct explanation.
 
Regards,
Jathushan Rajasegaran

**Comment by HopefulRational** [comment]
Thank for your reply. Eagerly awaiting a clarification

**Comment by ierosodin** [comment]
> @HopefulRational Hi, Thanks for your interest in our work.
> 
> 1. Our apologies for making the code, and figs bit complex. In the paper the `l+1` means the input to the `l+1`th layer, which is actually the output from layer `l` just after routing. Further, since the routing is iterative, this softmax is applied inside `update_routing` function on the output of the routing at each iteration.
> 2. Yes, I accept that the reshaping is too complicated. Yes, the `batch_size` dimension is changed to keep the multiplication with `logits`, without keeping the number of atoms in the dimension. but `votes_t_shape` and `r_t_shape` are simply transpose of themselves, and `r_t_shape` used later convert the votes back to early shape.
> 
> Again we are sorry for all the trouble you have to go through to understand the code, we will be releasing the pytorch implementation soon, which would be much easier to understand.
> 
> Please feel free to ask any questions or concerns you have.
> 
> Thank you very much.

Looking forward to the pytorch implementation!

**Comment by HopefulRational** [comment]
@ierosodin Hi. I have put a quick and dirty PyTorch implementation of DeepCaps as a .py file.You may have a look at that:
https://github.com/HopefulRational/DeepCaps-PyTorch
It is being trained on transformed FashionMNIST and tested of FashionMNIST.

**Comment by brjathu** [comment]
Hi,

Thank you very much, it seems fine for me, Did you get a chance to run it
on CIFAR10?

Regards,
R.Jathushan

**Comment by HopefulRational** [comment]
@ierosodin Hi. Thanks for liking it. Hope the bare-bones code was helpful :)
@brjathu Hi. May be i will just run on untransformed FashionMNIST once.

**Comment by brjathu** [comment]
@HopefulRational 
Hi, I sort of fixed the `update_routing` function. 
please check and Thank you very much for finding this bug.

**Comment by HopefulRational** [comment]
@brjathu Hi. Is this effecting the accuracies?

**Comment by brjathu** [comment]
I haven't tested it yet, I will run it today and let you know.

Regards,
R.Jathushan

**Comment by HopefulRational** [comment]
@brjathu Thanks.

**Comment by HopefulRational** [comment]
@brjathu by the way, I will go through the code in the evening coz i m little busy now.

**Comment by brjathu** [comment]
Test accuracy on cifar10 - 91.49%

Almost same as earlier one.

Regards,
R.Jathushan

**Comment by ierosodin** [comment]
Hi @brjathu ,
I still have a few doubts.
I'm wondering what's the meaning of "localized voting" you have mentioned in the paper.
For example, for each capsule in capsule layer l+1, what are the capsules it routes in capsule layer l? Every capsule in layer l across all of the location? or just in a limited kernel.
Thanks!

Sincerely,
Tony

**Comment by brjathu** [comment]
Hi,

Localized routing comes because, with a 3*3 kernel a nine local capsules
are routed together.

Regards,
R.Jathushan
